accepted/future-releases/spread-collections/performance.md

```diff
@@ -0,0 +1,400 @@
+### Iterable.forEach()
+
+This also works with any Iterable:
+
+```dart
+var temp = to;
+from.forEach((element) {
+  // Add element to result...
+});
+temp = null;
+```
+
+The shenanigans around `temp` are to ensure that the collection being spread
+can't hang onto the closure after the spread is complete, invoke it, and cause
+weird things to happen.
+
+Since it uses [internal iteration][internal], the collection being spread has
+maximum freedom to traverse itself efficiently. The flip side is that if we ever
+want to support `break` or other ways to interrupt a spread, internal iteration
+will make that *very* painful.
+
+Performance may be good, because it doesn't need to do any allocation.  On the
+other hand, it involves creating a closure which accesses a free variable. That
+can be slow on some implementations.
+
+[internal]: https://journal.stuffwithstuff.com/2013/01/13/iteration-inside-and-out/
+
+### List.[]
+
+If we know the object being spread happens to be a List, and not just an
+Iterable, we can use the List subscript API:
+
+```dart
+var length = from.length;
+for (var i = 0; i < length; i++) {
+  var element = from[i];
+  // Add element to result...
+}
+```
+
+This obviously doesn't work if the object isn't a List. Objects like the result
+of `.where()` and `.map()` are Iterables, but not Lists. But if it does work, it
+has a lot of promise. `.length` returns a simple integer and requires no
+allocation. For typical List classes, including the built in one, `[]` is simple
+and efficient: just an addition and a dereference.
+
+If it's fast enough for C, it's probably fast enough for everyone.
+
+### Runtime hybrid
+
+We could theoretically get the best of both worlds by checking at runtime to see
+if the Iterable is a List and choosing the API based on that:
+
+```dart
+if (from is List<String>) {
+  var length = from.length;
+  for (var i = 0; i < length; i++) {
+    var element = from[i];
+    // Add element to result...
+  }
+} else {
+  var iterator = from.iterator;
+  while (!iterator.moveNext()) {
+    var element = iterator.current;
+    // Add element to result...
+  }
+}
+```
+
+### Evaluation
+
+Here's what I found:
+
+*   `Iterable.forEach()` 1.3x faster than `Iterable.iterator` on AoT, 2.9x on
+    dart2js, but unfortunately 0.6x on the JIT VM.
+
+    I think the performance is too divergent to feel comfortable betting on
+    this. Also, I take some points away from this approach because using an
+    internal iterator may paint us into a corner with future control flow
+    changes. This isn't good enough to win out past that, even ignoring the poor
+    JIT performance.
+
+*   `List.[]` is much faster when the collection is empty: 3x Aot, 3x dart2js
+    and 10x on the JIT (!). This is probably showing the benefit of not
+    allocating the iterator. It's somewhat faster for non-empty collections too:
+    1.7x AoT, 1.2x dart2js, 1.3x JIT.
+
+    This looks pretty nice. I think empty collections will happen fairly
```

Comment:
If we allow the iteration to do `isEmpty` (or `isNotEmpty` or `length`) on the spread collection, then we can special-case the empty collection too.

Reply:
I did try that out but then realized this isn't a safe operation. If you call `isEmpty` on a WhereIterable before iterating it, for example, it will invoke the filter closure for the elements multiple times, which may have side effects.